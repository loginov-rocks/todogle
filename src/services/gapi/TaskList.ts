export default interface TaskList {
  kind: 'tasks#taskLis';
  id: string;
  etag: string;
  title: string;
  updated: string;
  selfLink: string;
}
