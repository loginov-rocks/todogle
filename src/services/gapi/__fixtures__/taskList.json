{
  "kind": "tasks#taskList",
  "id": "RUVUM3JvcDUxTUNHcDUtdw",
  "etag": "\"MTA0MzI0OTM1Nw\"",
  "title": "The best task list",
  "updated": "2020-07-11T14:01:56.445Z",
  "selfLink": "https://www.googleapis.com/tasks/v1/users/@me/lists/RUVUM3JvcDUxTUNHcDUtdw"
}
