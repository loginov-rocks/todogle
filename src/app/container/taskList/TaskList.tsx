import { Button, CircularProgress } from '@material-ui/core';
import * as React from 'react';

import gapi from '../../../services/gapi';
import { TaskListResource } from '../../../services/gapi/TaskListResource';
import { TaskResource } from '../../../services/gapi/TaskResource';

import CreateTask from './createTask/CreateTask';
import Task from './task/Task';

interface Props {
  onDelete: (id: string) => void;
  taskList: TaskListResource;
}

const TaskList = ({ onDelete, taskList }: Props) => {
  const [areTasksLoaded, setAreTasksLoaded] = React.useState(false);
  const [tasks, setTasks] = React.useState<TaskResource[]>([]);

  React.useEffect(() => {
    setAreTasksLoaded(false);

    gapi.getTasks(taskList.id)
      .then(tasks => {
        setAreTasksLoaded(true);
        setTasks(tasks);
      });
  }, [taskList.id]);

  const handleDelete = () => {
    gapi.deleteTaskList(taskList.id)
      .then(() => onDelete(taskList.id));
  };

  const handleTaskCreate = (task: TaskResource) => {
    setTasks(tasks.concat([task]));
  };

  const handleTaskDelete = (taskId: string) => {
    setTasks(tasks.filter(task => task.id !== taskId));
  };

  return (
    <>
      <b>{taskList.title}</b>
      <Button onClick={handleDelete} variant="contained">Delete</Button>

      {areTasksLoaded ? (
        <>
          <CreateTask
            onCreate={handleTaskCreate}
            taskListId={taskList.id}
          />

          {tasks.map(task => (
            <Task
              key={task.id}
              onDelete={handleTaskDelete}
              task={task}
              taskList={taskList}
            />
          ))}
        </>
      ) : (
        <CircularProgress />
      )}
    </>
  );
};

export default TaskList;
