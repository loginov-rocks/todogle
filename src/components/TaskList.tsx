import { Button, CircularProgress } from '@material-ui/core';
import * as React from 'react';

import CreateTask from './CreateTask';
import gapi from '../services/gapi';
import { TaskListResource } from '../services/gapi/TaskListResource';
import { TaskResource } from '../services/gapi/TaskResource';
import Task from './Task';

interface Props {
  onDelete: (id: string) => void;
  taskList: TaskListResource;
}

interface State {
  areTasksLoaded: boolean,
  tasks: TaskResource[];
}

export default class TaskList extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      areTasksLoaded: false,
      tasks: [],
    };

    this.handleDelete = this.handleDelete.bind(this);
    this.handleTaskCreate = this.handleTaskCreate.bind(this);
    this.handleTaskDelete = this.handleTaskDelete.bind(this);
  }

  componentDidMount() {
    this.updateTasks();
  }

  componentDidUpdate(prevProps: Readonly<Props>) {
    const { taskList } = this.props;

    if (taskList.id !== prevProps.taskList.id) {
      this.setState({
        areTasksLoaded: false,
      });

      this.updateTasks();
    }
  }

  updateTasks() {
    const { taskList } = this.props;

    gapi.getTasks(taskList.id)
      .then(tasks => {
        this.setState({
          areTasksLoaded: true,
          tasks,
        });
      });
  }

  handleDelete() {
    const { onDelete, taskList } = this.props;

    gapi.deleteTaskList(taskList.id)
      .then(() => onDelete(taskList.id));
  }

  handleTaskCreate(task: TaskResource) {
    const { tasks } = this.state;

    this.setState({
      tasks: tasks.concat([task]),
    });
  }

  handleTaskDelete(taskId: string) {
    const { tasks } = this.state;

    this.setState({
      tasks: tasks.filter(task => task.id !== taskId),
    });
  }

  render() {
    const { taskList } = this.props;
    const { areTasksLoaded, tasks } = this.state;

    return (
      <>
        <b>{taskList.title}</b>
        <Button onClick={this.handleDelete} variant="contained">Delete</Button>

        {areTasksLoaded ? (
          <>
            <CreateTask
              onCreate={this.handleTaskCreate}
              taskListId={taskList.id}
            />

            {tasks.map(task => (
              <Task
                key={task.id}
                onDelete={this.handleTaskDelete}
                task={task}
                taskList={taskList}
              />
            ))}
          </>
        ) : (
          <CircularProgress />
        )}
      </>
    );
  }
}
