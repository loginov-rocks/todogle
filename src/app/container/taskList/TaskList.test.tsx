import { shallow } from 'enzyme';
import * as React from 'react';

import taskList from '../../../services/gapi/__fixtures__/taskList.json';

import TaskList from './TaskList';

it('matches snapshot', () => {
  const wrapper = shallow(
    <TaskList onDelete={() => undefined} taskList={taskList} />,
  );

  expect(wrapper).toMatchSnapshot();
});
