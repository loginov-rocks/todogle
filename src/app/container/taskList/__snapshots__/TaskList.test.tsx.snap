// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<Fragment>
  <b>
    The best task list
  </b>
  <WithStyles(ForwardRef(Button))
    onClick={[Function]}
    variant="contained"
  >
    Delete
  </WithStyles(ForwardRef(Button))>
  <WithStyles(ForwardRef(CircularProgress)) />
</Fragment>
`;
